agent/service_manager.go

```diff
@@ -485,36 +476,34 @@ func makeConfigRequest(agent *Agent, registration *serviceRegistration) *structs
 	return req
 }
 
-// mergeServiceConfig returns the final effective config for the watched service,
-// including the latest known global defaults from the servers.
-//
-// NOTE: this is called while holding the Agent.stateLock
```

Comment:
This function is called from two places, only one of which is handling the `Agent.stateLock`. The other caller (`handleUpdate`) has a similar comment saying:

> the caller must NOT hold the Agent.stateLock!